### PR TITLE
weights option

### PR DESCRIPTION
**Pull request by JAMSADIQ**
Adding options of weight to compute weighted KDE.

### CODE REVIEW

**Comment by tdent** [comment]
Do you not also want to provide the option for AdaptiveBwKDE in the adaptive_kde module?

File: popde/density_estimate.py [hunk 1]
```diff
@@ -74,6 +77,18 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
         if dim_names is not None:
             self.check_dimensionality()
 
+        self.weights = weights
+        if self.weights is not None:
+            #check the array
+            self._weights = atleast_1d(weights).astype(float)
```
Comment (tdent) [inline popde/density_estimate.py:83]: I am not seeing what this `_weights` attribute is for or why we have both this and self.weights.

File: popde/density_estimate.py [hunk 2]
```diff
@@ -74,6 +77,18 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
...
+            #normalize weights [sum(weights)=1]
+            self._weights /= sum(self._weights)
+            if self.weights.ndim != 1:
+                raise ValueError("weights should be one-dimensional.")
+            if len(self.weights) != self.data.shape[1]:
+                raise ValueError("weights should be of length of input data")
```
Comment (tdent) [inline popde/density_estimate.py:89]: These checks should be performed before doing the normalization

File: popde/density_estimate.py [hunk 3]
```diff
@@ -74,6 +77,18 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
         if dim_names is not None:
             self.check_dimensionality()
 
+        self.weights = weights
+        if self.weights is not None:
+            #check the array
+            self._weights = atleast_1d(weights).astype(float)
+            #normalize weights [sum(weights)=1]
+            self._weights /= sum(self._weights)
```
Comment (tdent) [inline popde/density_estimate.py:85]:
```suggestion
            # Normalize to sum to 1
            self._weights /= self._weights.sum()
```

File: popde/density_estimate.py [hunk 4]
```diff
@@ -74,6 +77,18 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
...
+            if self.weights.ndim != 1:
+                raise ValueError("weights should be one-dimensional.")
+            if len(self.weights) != self.data.shape[1]:
+                raise ValueError("weights should be of length of input data")
+
```
Comment (tdent) [inline popde/density_estimate.py:90]:
```suggestion
```

File: popde/density_estimate.py (hunk 1 above)
Comment (JAMSADIQ) [reply popde/density_estimate.py]: fixed it.

File: popde/density_estimate.py (hunk 2 above)
Comment (JAMSADIQ) [reply popde/density_estimate.py:87]: rearrange

File: popde/density_estimate.py (hunk 4 above)
Comment (JAMSADIQ) [reply popde/density_estimate.py]: done

File: popde/density_estimate.py [hunk 5]
```diff
@@ -74,6 +77,17 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
...
 
+        self.weights = weights
+        if self.weights is not None:
+            # Check the array
+            self.weights = atleast_1d(weights).astype(float)
+            if self.weights.ndim != 1:
+                raise ValueError("weights should be one-dimensional.")
+            if len(self.weights) != self.data.shape[1]:
+                raise ValueError("weights should be of length of input data")
+            # Normalize to sum to  1
```
Comment (tdent) [inline popde/density_estimate.py:88]:
```suggestion
            # Normalize to sum to 1
```

File: popde/density_estimate.py [hunk 6]
```diff
@@ -31,6 +31,9 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
             backend : String, Processing method to do KDE calculation
```
Comment (tdent) [inline popde/density_estimate.py:31]:
```suggestion
            backend : string, Processing method to do KDE calculation
```

File: popde/density_estimate.py [hunk 7]
```diff
@@ -31,6 +31,9 @@ def __init__(self, data, input_transf=None, stdize=False, rescale=None,
...
             bandwidth : Float or array of float, bandwidth of kernels used for smoothing
             dim_names : Sequence of dimension names, e.g. ('m1', 'z', 'chi_eff')
+            weights : array_like, optional None (default)
+                      weights of datapoints so  must be the same shape 
+                      as n_samples. If None (default), the samples are assumed to be                        equally weighted
```
Comment (tdent) [inline popde/density_estimate.py:36]:
```suggestion
            weights : array_like, weights for data points ; if None, the samples are weighted equally
```
We don't need to state the default in the docstring if it is obvious

**Review by tdent — CHANGES_REQUESTED** [review]
Minor edits to comments but otherwise looks OK

File: popde/density_estimate.py [hunk 8]
```diff
@@ -313,7 +326,7 @@ def plot_2d_contour(self, dim1, dim2, slice_dims=None, slice_values=None, num_po
 
         # Create the contour plot
         zz = z.reshape(xx.shape)
-        fig = utils_plot.simple2Dplot(xx, yy, zz, xlabel=dim1, ylabel=dim2, title=f'KDE sliced along {slice_dims} at {slice_values})')
+        fig = utils_plot.simple2Dplot(xx, yy, zz, xlabel=dim1, ylabel=dim2, title=f'KDE sliced along {slice_dims} at {slice_values})', show_plot=True)
```
Comment (tdent) [inline popde/density_estimate.py:329]: why has this been changed?

Comment (JAMSADIQ) [reply popde/density_estimate.py]: I removed it. I was testing the code for examples and mistakenly commit this. Fixed it now

**Review by tdent — APPROVED** [review]

**Pull request merged**